framework/include/utils/RankTwoTensorImplementation.h

```diff
@@ -949,137 +965,32 @@ RankTwoTensor::symmetricEigenvaluesEigenvectors(std::vector<Real> & eigvals,
 
 template <>
 void
-ADRankTwoTensor::symmetricEigenvaluesEigenvectors(std::vector<DualReal> & eigvals,
+ADRankTwoTensor::symmetricEigenvaluesEigenvectors(std::vector<ADReal> & eigvals,
                                                   ADRankTwoTensor & eigvecs) const
 {
-  const Real eps = libMesh::TOLERANCE * libMesh::TOLERANCE;
-
-  eigvals.resize(N);
-  ADRankTwoTensor D, Q, R;
-  this->hessenberg(D, eigvecs);
-
-  unsigned int iter = 0;
-  for (unsigned m = N - 1; m > 0; m--)
-    do
-    {
-      iter++;
-      auto shift = D(m, m);
-      D.addIa(-shift);
-      D.QR(Q, R, m + 1);
-      D = R * Q;
-      D.addIa(shift);
-      eigvecs = eigvecs * Q;
-    } while (std::abs(D(m, m - 1)) > eps);
-
-  for (unsigned int i = 0; i < N; i++)
-    eigvals[i] = D(i, i);
-
-  // Sort eigenvalues and corresponding vectors.
-  for (unsigned int i = 0; i < N - 1; i++)
-  {
-    unsigned int k = i;
-    auto p = eigvals[i];
-    for (unsigned int j = i + 1; j < N; j++)
-      if (eigvals[j] < p)
-      {
-        k = j;
-        p = eigvals[j];
-      }
-    if (k != i)
+  typedef Eigen::Matrix<ADReal, N, N, Eigen::DontAlign> RankTwoMatrix;
+  RankTwoMatrix self;
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = i; j < N; ++j)
     {
-      eigvals[k] = eigvals[i];
-      eigvals[i] = p;
-      for (unsigned int j = 0; j < N; j++)
-      {
-        p = eigvecs(j, i);
-        eigvecs(j, i) = eigvecs(j, k);
-        eigvecs(j, k) = p;
-      }
+      auto & v = self(j, i);
+      v = (*this)(i, j);
+      if (i != j && MooseUtils::absoluteFuzzyEqual(v, 0.0))
+        v.value() = 0.0;
```

Comment:
I don't know 100% what's going on here in this class, but in this particular line of code I would clear the derivatives too, so just
```suggestion
        v = 0.0;
```
In general I am frightened of any calls to `.value()` or `.derivatives()` (unless the latter call is for initial seeding)

Reply:
Don't we lose all dual numbers by clearing derivatives?

@dschwen Do you remember why we did this `v.value() = 0.0` here? I thought a value of zero doesn't perform well in degenerate cases. Perhaps I had the wrong impression. Anyways, I think adding a brief comment here would be a good idea.

Reply:
> Don't we lose all dual numbers by clearing derivatives?

I don't understand this statement. `v` is still a `DualNumber` no matter what you do here. I was just commenting that it feels very concerning to suddenly assign its value to zero and not do anything with the derivatives, which mathematically to me feels like changing the definition of f(x) without changing the definition of f'(x)

Reply:
The assumption is that for a continuous differentiable function the derivative would be very similar at zero than at epsilon away from zero.

Reply:
I certainly agree with that but when I see manual manipulation like this it sure doesn't look like a continuous differentiable function. If you are leaving the derivatives unaltered, why are you manually setting to `0.` when you are somewhere close to 0? Is there optimization somewhere else that only happens when the value is "exactly" `0.`?

Reply:
e.g. why can't you just leave the `.value()` as "something close to zero" as opposed to "exactly" 0

Reply:
Yes. I do not remember the specifics, but the Eigen decomposition fails or returns NaN if off diagonal values are close to but not exactly zero.

Reply:
I can certainly re-investigate, but I spent some effort looking at the performance of real-life simulations with this PR. I think I'd have to redo that analysis if make this change (or the two variations of the change) you are asking for. As it is, the PR shows favorable performance compared to devel.

Reply:
Ok, if you get "better" behavior by exactly setting to zero, then it makes sense to me to do what you're doing, both with respect to the value and derivatives

Reply:
It would be nice to have some comment about the motivation for this manual setting to zero, e.g. what you said about failed Eigen decompositions. For me as the reviewer and possibly future people who come across this code, this manual setting to 0 seems to come out of nowhere.